src/shell/info.c

```diff
@@ -45,33 +43,80 @@ static int lookup_job_info_get (flux_future_t *f,
         shell_log_error ("failed to unwrap J: %s", error.text);
         return -1;
     }
-    if (flux_rpc_get_unpack (f, "{s:s}", "R", R) < 0)
-        goto error;
     return 0;
 error:
     shell_log_error ("job-info: %s", future_strerror (f, errno));
     return -1;
 }
 
-/* Fetch R and J from the job-info service.
+/* Fetch J from the job-info service.
  * Return future on success or NULL on failure (and log error).
  */
-static flux_future_t *lookup_job_info (flux_t *h, flux_jobid_t jobid)
+static flux_future_t *lookup_jobspec (flux_t *h, flux_jobid_t jobid)
 {
     flux_future_t *f;
     f = flux_rpc_pack (h,
                        "job-info.lookup",
                        FLUX_NODEID_ANY,
                        0,
-                       "{s:I s:[ss] s:i}",
+                       "{s:I s:[s] s:i}",
                        "id", jobid,
-                       "keys", "R", "J",
+                       "keys", "J",
                        "flags", 0);
     if (!f)
         shell_log_error ("error sending job-info request");
     return f;
 }
 
+static void resource_watch_update (struct shell_info *info)
+{
+    flux_future_t *f = info->R_watch_future;
+    json_t *R = NULL;
+    json_t *tmp_R;
+    rcalc_t *rcalc = NULL;
+    rcalc_t *tmp_rcalc;
+
+    if (flux_rpc_get_unpack (f, "{s:o}", "R", &R) < 0) {
+        shell_log_errno ("error getting R from job-info watch response");
+        goto out;
+    }
+    if (!(rcalc = rcalc_create_json (R))) {
+        shell_log_error ("error decoding R");
+        goto out;
+    }
+    /*  Swap previous and updated R, rcalc:
+     */
+    tmp_R = info->R;
+    info->R = json_incref (R);
+    R = tmp_R;
+
+    tmp_rcalc = info->rcalc;
+    info->rcalc = rcalc;
+    rcalc = tmp_rcalc;
+out:
+    rcalc_destroy (rcalc);
+    json_decref (R);
+    flux_future_reset (f);
```

Comment:
The error handling looks incomplete here.  `flux_future_reset()` is called when `flux_rpc_get_unpack()` fails, but the future won't ever be fulfilled again.  Also no error is propagated to `R_update_cb()` which proceeds to call the plugstack.  Is that a problem?

Reply:
> flux_future_reset() is called when flux_rpc_get_unpack() fails, but the future won't ever be fulfilled again

This allows a job to continue running if there's an error updating the shell's internal R. It doesn't seem like it should be fatal -- perhaps there will not even be any side effects. I'm not sure why the future would never be fulfilled again. If R is updated again, presumably the future would be fulfilled again?

> Also no error is propagated to R_update_cb() which proceeds to call the plugstack. Is that a problem?

Yeah, I suppose that should be fixed. The plugstack would be called without an R update which should have no effect but it is extra work for no reason at the worst case.

Reply:
A streaming rpc is terminated by an error response so no further responses are possible. But the point is somewhat moot if nothing is waiting on it and it should not be fatal.

We will need to revisit once we support reconnecting after instance restart.

Reply:
An unpack error could also occur here and is the presumed cause of any error here. In that case I assume the streaming RPC would not be terminated? The job cannot end until the shell exits so `ENODATA` at least is not expected.

Reply:
I did update the function to return -1 on error and if there is an error on the first response (the one fetched synchronously) then that is indeed a fatal error. So thanks for catching that it is was actually a pretty severe bug!

Reply:
I pushed a fixup commit for this issue. `resource_watch_update()` now returns -1 on failure. If this function fails on the first response, then that is a fatal error, o/w the error is ignored (though the issue is logged).

Reply:
Is there a potential double-free  on `R` if `rcalc_create_json()` fails?  Could the bottom half of this function be simplified to
```c
    json_decref (info->R);
    info->R = json_incref (R);
    rcalc_destroy (info->rcalc);
    info->rcalc = rcalc;
    rc = 0;
out:
    flux_future_reset (f);
    return rc;
```

Reply:
I think my comment about the double free and possible simplification in the shell's `resource_watch_update()` was overlooked?